src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp

```diff
@@ -0,0 +1,311 @@
+	} else {
+		// no escape provided: try standard/common escapes
+		quoterule_candidates = {QuoteRule::QUOTES_RFC, QuoteRule::QUOTES_OTHER, QuoteRule::NO_QUOTES};
+	}
+}
+
+void CSVSniffer::GenerateStateMachineSearchSpace(vector<unique_ptr<CSVStateMachine>> &csv_state_machines,
+                                                 const vector<char> &delim_candidates,
+                                                 const vector<QuoteRule> &quoterule_candidates,
+                                                 const vector<vector<char>> &quote_candidates_map,
+                                                 const vector<vector<char>> &escape_candidates_map) {
+	// Generate state machines for all option combinations
+	for (auto quoterule : quoterule_candidates) {
+		const auto &quote_candidates = quote_candidates_map[static_cast<uint8_t>(quoterule)];
+		for (const auto &quote : quote_candidates) {
+			for (const auto &delim : delim_candidates) {
+				const auto &escape_candidates = escape_candidates_map[static_cast<uint8_t>(quoterule)];
+				for (const auto &escape : escape_candidates) {
+					D_ASSERT(buffer_manager);
+					csv_state_machines.emplace_back(
+					    make_uniq<CSVStateMachine>(options, quote, escape, delim, buffer_manager,
+					                               buffer_manager->context.client_data->state_machine_cache));
+				}
+			}
+		}
+	}
+}
+
+void CSVSniffer::AnalyzeDialectCandidate(unique_ptr<CSVStateMachine> state_machine, idx_t &rows_read,
+                                         idx_t &best_consistent_rows, idx_t &prev_padding_count,
+                                         idx_t prev_column_count) {
+	vector<idx_t> sniffed_column_counts(options.sample_chunk_size);
+	state_machine->csv_buffer_iterator.Process<SniffDialect>(*state_machine, sniffed_column_counts);
+	idx_t start_row = options.dialect_options.skip_rows;
+	idx_t consistent_rows = 0;
+	idx_t num_cols = sniffed_column_counts.empty() ? 0 : sniffed_column_counts[0];
+	idx_t padding_count = 0;
+	bool allow_padding = options.null_padding;
+	if (sniffed_column_counts.size() > rows_read) {
+		rows_read = sniffed_column_counts.size();
+	}
+	for (idx_t row = 0; row < sniffed_column_counts.size(); row++) {
+		if (sniffed_column_counts[row] == num_cols) {
+			consistent_rows++;
+		} else if (num_cols < sniffed_column_counts[row] && !options.skip_rows_set) {
+			// all rows up to this point will need padding
+			padding_count = 0;
+			// we use the maximum amount of num_cols that we find
+			num_cols = sniffed_column_counts[row];
+			start_row = row + options.dialect_options.skip_rows;
+			consistent_rows = 1;
+
+		} else if (num_cols >= sniffed_column_counts[row]) {
+			// we are missing some columns, we can parse this as long as we add padding
+			padding_count++;
+		}
+	}
+
+	if (num_cols < prev_column_count) {
+		// Early return if we have less columns than the previous chunk run
+		return;
+	}
```

Comment:
Is this for simplicity? If possible, does null-padding to the maximum column count add unnecessary complexity?

Reply:
In case padding is not accepted?

Reply:
Oh yea, true, I missed that.